### PR TITLE
Compat between sbt.Logger and ProcessLogger

### PR DESCRIPTION
**Pull request by MasseGuillaume**

### CODE REVIEW

File: src/main/scala-sbt-0.13/compat.scala [hunk 1]
```diff
@@ -10,6 +10,15 @@ object compat {
...
     def withName(n: String): ModuleID = _m copy (name = n)
   }
+
+  implicit class RichProcessBuilder(pb: ProcessBuilder) {
+    def lineStream_!(logger: ProcessLogger) = pb.lines_!(logger)
+    def lineStream_! : Stream[String]       = pb.lines_!
+  }
+
+  implicit def promoteProcessLogger(logger: sbt.Logger): ProcessLogger =
+    ProcessLogger(msg => logger.info(msg), msg => logger.error(msg))
```
Comment (dwijnand) [inline src/main/scala-sbt-0.13/compat.scala]: Hmm, why is this required? There is already an implicit lift from Logger to ProcessLogger in sbt 0.13:

https://github.com/sbt/sbt/blob/v0.13.16/util/log/src/main/scala/sbt/Logger.scala#L58

Comment (MasseGuillaume) [reply src/main/scala-sbt-0.13/compat.scala]: sbt.Logger is not available in sbt 1.0.

Comment (dwijnand) [reply src/main/scala-sbt-0.13/compat.scala]: In sbt 1 sbt.Logger still exists. It's defined at sbt.util.Logger (https://github.com/sbt/util/blob/v1.0.2/internal/util-logging/src/main/scala/sbt/util/Logger.scala#L19) and then aliased as sbt.Logger (https://github.com/sbt/sbt/blob/v1.0.2/sbt/src/main/scala/Import.scala#L81-L82)

Comment (MasseGuillaume) [reply src/main/scala-sbt-0.13/compat.scala]: indeed, it's not needed you are right.

Comment (MasseGuillaume) [reply src/main/scala-sbt-0.13/compat.scala]: however, you need to  `import sbt.Logger._`. I think it would make sense to make it available here as well. This way, you only need to `import sbt.compat._` and it solves several issues with one import.

Comment (dwijnand) [reply src/main/scala-sbt-0.13/compat.scala]: Are you sure? Unless I'm missing something by being defined on the companion object it's in implicit scope, for both sbt 0.13 and sbt 1, so it's ok?

Btw I'm happy to find a way to include this is need be, I'm just not convinced of the need.

File: src/main/scala-sbt-0.13/compat.scala [hunk 2]
```diff
@@ -4,12 +4,21 @@ package sbt
...
+
 object compat {
   val scalaModuleInfo = SettingKey[Option[librarymanagement.ScalaModuleInfo]]("ivyScala")
 
   implicit class ModuleIdOps(val _m: ModuleID) extends AnyVal {
     def withName(n: String): ModuleID = _m copy (name = n)
   }
+
+  implicit class RichProcessBuilder(pb: ProcessBuilder) {
+    def lineStream_!(logger: ProcessLogger) = pb.lines_!(logger)
```
Comment (dwijnand) [inline src/main/scala-sbt-0.13/compat.scala:17]: Sorry Guillaume , could you also add a little comment explaining that these are here because `lines_!` is deprecated in Scala 2.11+?

Took me a while to go through the history of sbt 0.13/1.0, Scala 2.10/2.11/2.12 to understand why this is here :)

File: src/main/scala-sbt-0.13/compat.scala [hunk 3]
```diff
@@ -4,12 +4,21 @@ package sbt
...
+    def lineStream_! : Stream[String]       = pb.lines_!
+  }
+
+  implicit def absLog2PLog(log: AbstractLogger): ProcessLogger = sbt.Logger.absLog2PLog(log)
```
Comment (dwijnand) [inline src/main/scala-sbt-0.13/compat.scala:21]: My response is in a collapsed thread, with the new commit, so reposting here:

> however, you need to import sbt.Logger._. I think it would make sense to make it available here as well. This way, you only need to import sbt.compat._ and it solves several issues with one import.

Are you sure? Unless I'm missing something by being defined on the companion object it's in implicit scope, for both sbt 0.13 and sbt 1, so it's ok?

Btw I'm happy to find a way to include this is need be, I'm just not convinced of the need.

**Review by dwijnand — CHANGES_REQUESTED** [review]

File: src/main/scala-sbt-0.13/compat.scala (hunk 2 above)
Comment (MasseGuillaume) [reply src/main/scala-sbt-0.13/compat.scala:17]: Yeah. This is only related to 2.10 vs 2.12. If you use `lines_!` you get a deprecation warning in 2.12, but you can't use `lineStream_!` in 2.10 because it does not exists.

File: src/main/scala-sbt-0.13/compat.scala (hunk 3 above)
Comment (MasseGuillaume) [reply src/main/scala-sbt-0.13/compat.scala:21]: from scala-native:

```scala
import scala.sys.process._
class ComRunner(bin: File,
                envVars: Map[String, String],
                args: Seq[String],
                logger: Logger) {

  private[this] val runner = new Thread {
    override def run(): Unit = {
      val port = serverSocket.getLocalPort
      logger.info(s"Starting process '$bin' on port '$port'.")
      Process(bin.toString +: port.toString +: args, None, envVars.toSeq: _*) ! logger
    }
  }
```

```
[error]  found   : sbt.Logger
[error]  required: scala.sys.process.ProcessLogger
[error]       Process(bin.toString +: port.toString +: args, None, envVars.toSeq: _*) ! logger
```

Comment (dwijnand) [reply src/main/scala-sbt-0.13/compat.scala:21]: Ok now I see what the use case is. So you want to use scala.sys.process's Process with sbt's Logger.

To do that for both sbt 0.13 and sbt 1 code I think you'd need implicit conversions for both.

**Comment by dwijnand** [comment]
Closing as stale. I'm happy to review a refreshed version of this!
